### PR TITLE
Read additional module and theme folders from web.config, take 2

### LINKED ISSUE #5019: Ability to load extensions from different folder
@Piedone created:
https://orchard.codeplex.com/workitem/21190

It would be great to be able to load modules and themes from other folders than just the default Modules and Themes folder.

There is partial support for this today (IExtensionFolders) but the strings "~/Modules" and "~/Themes" are hard-coded in many places that prevent you from e.g. using templates from modules stored in a different folder than /Modules.

Being able to use arbitrary extension folders (defined in e.g. Web.config, HostComponents.config) would be a step forward. Still, since it's a web app operating with virtual paths you'd have to place your extensions under Orchard.Web (though you could do that into new folders).

Particularly Drupal lets you place your own extensions under a different folder than the built-in ones: https://www.drupal.org/node/2116781

Also see: https://orchard.codeplex.com/discussions/570620

### PR DESCRIPTION
**Pull request by qt1**
updated version of 
https://github.com/OrchardCMS/Orchard/pull/5357

Should resolve #5019, #5241, #5354

Additional comma separated lists can be defined in web.config (example included).
Default is added to the list in any case.
Includes trimming, should ignore empty strings and duplicates.

### CODE REVIEW

File: src/Orchard.Tests.Modules/Migrations/SchemaCommandGeneratorTests.cs [hunk 1]
```diff
@@ -118,6 +118,7 @@ public class SchemaCommandGeneratorTests {
         }
 
         public class StubFolders : IExtensionFolders {
+            public IEnumerable<string> Paths { get { return AvailableExtensions().Select(e => e.Location).Distinct(); } }
```
Comment (Piedone) [inline src/Orchard.Tests.Modules/Migrations/SchemaCommandGeneratorTests.cs]: This could go into an abstract base class instead of being copied over multiple times.

File: src/Orchard.Web/Web.config [hunk 2]
```diff
@@ -15,6 +15,12 @@
         <add key="webpages:Version" value="3.0.3"/>
         <add key="log4net.Config" value="Config\log4net.config"/>
         <add key="owin:AppStartup" value="Orchard.Owin.Startup, Orchard.Framework"/>
+    
+<!-- optional additional comma-separated list of folders for Modules or Themes -->
+        <add key="Modules" value="~/Ibn_Modules"/>
+<!--
+        <add key="Themes" value="~/Themes/MoreThemes" />
+-->
```
Comment (Piedone) [inline src/Orchard.Web/Web.config]: Sebastien also mentioned under the issue that it's probably best to have this config in the HostComponents.config file and consume it similarly how the services do it there.

File: src/Orchard/Environment/Configuration/ShellSettings.cs [hunk 3]
```diff
@@ -15,6 +15,15 @@ public class ShellSettings {
...
 
+        public string[] _globalPathPrefixes = { };
+        public string[] GlobalPathPrefixes { get { return _globalPathPrefixes; } }
+        private static string[] defaultGlobalPrefixes = { "~/Media" }; 
+
+        public void updateGlobalPathPrefixes(IEnumerable<string> locations)
+        {
+            _globalPathPrefixes = defaultGlobalPrefixes.Concat(locations).ToArray();
+        }
+
```
Comment (Piedone) [inline src/Orchard/Environment/Configuration/ShellSettings.cs]: All this seems off. Why in ShellSettings if it's not actually part of shell settings (nor is it persisted)? Names are unconventional too.

File: src/Orchard/Environment/Extensions/Folders/IExtensionFolders.cs [hunk 4]
```diff
@@ -4,5 +4,6 @@
 namespace Orchard.Environment.Extensions.Folders {
     public interface IExtensionFolders {
         IEnumerable<ExtensionDescriptor> AvailableExtensions();
+        IEnumerable<string> Paths { get; }
```
Comment (Piedone) [inline src/Orchard/Environment/Extensions/Folders/IExtensionFolders.cs]: I don't think it's good to have this property on the interface. IExtensionFolders is quite a high-level concept (though naming it "Folders" gives away some clues...) and I wouldn't surface the internally used paths like this (it doesn't really make sense for CoreModuleFolders e.g.). There can be simply a public property like this just on the module and theme folders.

File: src/Orchard/Environment/Extensions/Models/ExtensionDescriptor.cs [hunk 5]
```diff
@@ -12,6 +12,8 @@ public class ExtensionDescriptor {
         /// </summary>
         public string Id { get; set; }
 
+        public string VirtualPath { get { return Location + "/" + Id; } }
```
Comment (Piedone) [inline src/Orchard/Environment/Extensions/Models/ExtensionDescriptor.cs]: Is this really needed here? Also don't concatenate paths by hand, use VirtualPathUtility.Combine().

File: src/Orchard/Environment/OrchardStarter.cs [hunk 6]
```diff
@@ -174,6 +176,16 @@ public static class OrchardStarter {
             return container;
         }
 
+        /// <summary>
+        /// Get list of comma separated paths from web.config appSettings
+        /// Also return the default path
+        /// </summary>
+        static IEnumerable<string> GetConfigPaths(string key, string defaultPath)
+        {
+            char[] delim = new char[]{','};
```
Comment (Piedone) [inline src/Orchard/Environment/OrchardStarter.cs]: Use var, don't abbreviate variable names and format the code.

File: src/Orchard/FileSystems/Dependencies/DynamicModuleVirtualPathProvider.cs [hunk 7]
```diff
@@ -15,10 +18,16 @@ namespace Orchard.FileSystems.Dependencies {
     /// </summary>
     public class DynamicModuleVirtualPathProvider : VirtualPathProvider, ICustomVirtualPathProvider {
         private readonly IExtensionDependenciesManager _extensionDependenciesManager;
+        string[] ExtensionsVirtualPathPrefixes;
```
Comment (Piedone) [inline src/Orchard/FileSystems/Dependencies/DynamicModuleVirtualPathProvider.cs]: This is named unconventionally (should be private field).

File: src/Orchard/FileSystems/Dependencies/DynamicModuleVirtualPathProvider.cs [hunk 8]
```diff
@@ -15,10 +18,16 @@ namespace Orchard.FileSystems.Dependencies {
...
     public class DynamicModuleVirtualPathProvider : VirtualPathProvider, ICustomVirtualPathProvider {
         private readonly IExtensionDependenciesManager _extensionDependenciesManager;
+        string[] ExtensionsVirtualPathPrefixes;
 
-        public DynamicModuleVirtualPathProvider(IExtensionDependenciesManager extensionDependenciesManager) {
+        public DynamicModuleVirtualPathProvider(IExtensionDependenciesManager extensionDependenciesManager, IEnumerable<IExtensionFolders> folders) {
             _extensionDependenciesManager = extensionDependenciesManager;
             Logger = NullLogger.Instance;
+
+            ExtensionsVirtualPathPrefixes = folders
```
Comment (Piedone) [inline src/Orchard/FileSystems/Dependencies/DynamicModuleVirtualPathProvider.cs]: Better sould be to have a Lazy<T>.

File: src/Orchard/Utility/Extensions/HttpRequestExtensions.cs [hunk 9]
```diff
@@ -85,6 +85,10 @@ public static class HttpRequestExtensions {
                 return true;
             }
 
+            if (url.StartsWith("../")) {
```
Comment (Piedone) [inline src/Orchard/Utility/Extensions/HttpRequestExtensions.cs]: Hmm, how is this change related to the improvements?

File: src/Orchard.Tests.Modules/Migrations/SchemaCommandGeneratorTests.cs (hunk 1 above)
Comment (qt1) [inline src/Orchard.Tests.Modules/Migrations/SchemaCommandGeneratorTests.cs]: Indeed. However the class itself is repeated several times, so if touching this It may make sense to reduce that redundancy on the first place.. 
Since all the repetitions are part of various tests I would leave this to after we verify all else works..

File: src/Orchard.Web/Web.config (hunk 2 above)
Comment (qt1) [inline src/Orchard.Web/Web.config]: web.config has two benefits:
 1. it is definition and not implementation
 2. web.config is better known and its syntax it more accessible (users do not need to understand the implementation details).

My personal preference is to keep it in web.config.

File: src/Orchard/Environment/Extensions/Folders/IExtensionFolders.cs (hunk 4 above)
Comment (qt1) [inline src/Orchard/Environment/Extensions/Folders/IExtensionFolders.cs]: Modules and themes settings may change according to tenant.
There are several lists that are used' quit inconsistently around the code which include (If I am not missing any):
List of view lookup folders, list of dependency folders (excluding core), list of folders that should not be translated in tenant href.

It is possible that the best solution would be to define all such lists in some tenant-related settings object.

Where would you suggest?

File: src/Orchard/Environment/Extensions/Models/ExtensionDescriptor.cs (hunk 5 above)
Comment (qt1) [inline src/Orchard/Environment/Extensions/Models/ExtensionDescriptor.cs]: Actually there are several locations in the code where it is written 'VirtualPath = Location + "/" + Id'. and than VirtualPath is used.

Having explicit VirtualPath gives an explicit calculation instead of leaving the calculation to each implementation, it can reduced repetition and it can also be calculated once and used many times.

File: src/Orchard/Utility/Extensions/HttpRequestExtensions.cs (hunk 9 above)
Comment (qt1) [inline src/Orchard/Utility/Extensions/HttpRequestExtensions.cs]: This is just something to get some test code working. It may go to a different PR.

File: src/Orchard/Environment/Extensions/Folders/IExtensionFolders.cs (hunk 4 above)
Comment (Piedone) [inline src/Orchard/Environment/Extensions/Folders/IExtensionFolders.cs]: I'm under the impression you replied to this comment of mine: https://github.com/OrchardCMS/Orchard/pull/5604/files#r36408832

I think that such paths (if we're talking about path prefixes where actual extensions were loaded from) should be store on an app-level (not tenant-level): it's easier and since we don't talk about tens of thousands of possible locations, all different per tenants here I don't see much value in maintaining such a list for faster lookup (my impression is that this was the aim).

File: src/Orchard.Web/Web.config (hunk 2 above)
Comment (Piedone) [inline src/Orchard.Web/Web.config]: Good points, seems better to me now too.

File: src/Orchard/Environment/OrchardStarter.cs [hunk 10]
```diff
@@ -174,6 +176,16 @@ public static class OrchardStarter {
...
         }
 
+        /// <summary>
+        /// Get list of comma separated paths from web.config appSettings
+        /// Also return the default path
+        /// </summary>
+        static IEnumerable<string> GetConfigPaths(string key, string defaultPath)
+        {
+            char[] delim = new char[]{','};
+            return (WebConfigurationManager.AppSettings[key] ?? "").Split(delim, StringSplitOptions.RemoveEmptyEntries).Concat(new string[]{defaultPath}).Select(s => s.Trim()).Distinct(StringComparer.OrdinalIgnoreCase);
```
Comment (Piedone) [inline src/Orchard/Environment/OrchardStarter.cs]: Use ConfgurationManager, or better, IAppConfigurationAccessor if possible. The advantage of ConfigurationManager is that it's not ASP.NET-coupled, could be even used when Orchard is run as an exe with [Orchard App Host](http://orchardapphost.codeplex.com/). IAppConfigurationAccessor is of course also flexible in implementation (and by default uses ConfigurationManager)

File: src/Orchard/Environment/Extensions/Folders/IExtensionFolders.cs (hunk 4 above)
Comment (qt1) [inline src/Orchard/Environment/Extensions/Folders/IExtensionFolders.cs]: Make sense. 

At the moment the only location path are actually stored at is IExtensionFolders (everything else is hard coded).
I am not that familiar with Orchard classes so I'll appreciate if you could suggest a class for storing the path lists.
For the sake of efficiency I would store just path lists and not anything more fancy.

File: src/Orchard/Environment/Extensions/Models/ExtensionDescriptor.cs (hunk 5 above)
Comment (Piedone) [inline src/Orchard/Environment/Extensions/Models/ExtensionDescriptor.cs]: Fair enough. The concatenation is still not right, even if this was the case before.

**Comment by Piedone** [comment]
I made some quick tests and it appears to be working very well.

One slightly related problem I noticed that for themes put under ~/OtherThemes Theme.png is not properly retrieved: e.g. http://localhost:30322/OrchardLocal/OtherThemes/Lombiq.PrettyGoodBootstrapBaseTheme/Theme.png will result in a 404 (and the same of course for other themes put there too) while other static files work (e.g. http://localhost:30322/OrchardLocal/OtherThemes/Lombiq.PrettyGoodBootstrapBaseTheme/Styles/site.css). This is because the original Themes folder contains a Web.config that allows to serve specifically these files. This may be something to document (that you need to copy it over if your theme folder is not under ~/Themes).

**Comment by qt1** [comment]
Thanks for testing,

Indeed a theme that are not under ~/Themes should have its own web.config and I would recommend its own project file too.

It should also be noted that if a project depth in the tree is changed than some of its references may need to be updaed. (for example moving ~/Modules/MyModule to ~/CustomerA/Modules/MyModule)

**Comment by qt1** [comment]
Updated as per comments from reviewers.

File: src/Orchard/Localization/Services/DefaultLocalizedStringManager.cs [hunk 11]
```diff
@@ -17,8 +17,8 @@ public class DefaultLocalizedStringManager : ILocalizedStringManager {
         private readonly ISignals _signals;
 
         const string CoreLocalizationFilePathFormat = "~/Core/App_Data/Localization/{0}/orchard.core.po";
-        const string ModulesLocalizationFilePathFormat = "~/Modules/{0}/App_Data/Localization/{1}/orchard.module.po";
-        const string ThemesLocalizationFilePathFormat = "~/Themes/{0}/App_Data/Localization/{1}/orchard.theme.po";
+        const string ModulesLocalizationFilePathFormat = "{0}/App_Data/Localization/{1}/orchard.module.po";
```
Comment (sebastienros) [inline src/Orchard/Localization/Services/DefaultLocalizedStringManager.cs]: We need to test, even manually, that this is not breaking localization

File: src/Orchard/Mvc/ViewEngines/IViewEngineProvider.cs [hunk 12]
```diff
@@ -8,6 +8,7 @@ public class CreateThemeViewEngineParams {
 
     public class CreateModulesViewEngineParams {
         public IEnumerable<string> VirtualPaths { get; set; }
+        public IEnumerable<string> ModuleLocations { get; set; }
```
Comment (sebastienros) [inline src/Orchard/Mvc/ViewEngines/IViewEngineProvider.cs]: This is Extensions, not just modules

File: src/Orchard/Mvc/ViewEngines/ThemeAwareness/ThemeAwareViewEngine.cs [hunk 13]
```diff
@@ -93,12 +94,17 @@ public class ThemeAwareViewEngine : IThemeAwareViewEngine {
...
-                var allLocations = enabledModules.Concat(enabledModules)
-                    .Select(fd => fd.Extension.Location + "/" + fd.Extension.Id)
-                    .Distinct(StringComparer.OrdinalIgnoreCase)
+                var moduleVirtualPaths = enabledModules
+                    .Select(fd => fd.Extension.VirtualPath)
+                    .Distinct(StringComparer.OrdinalIgnoreCase) // is Distinct guaranty to keep order?
                     .ToList();
 
-                var moduleParams = new CreateModulesViewEngineParams { VirtualPaths = allLocations };
+                var moduleLocations = enabledModules
```
Comment (sebastienros) [inline src/Orchard/Mvc/ViewEngines/ThemeAwareness/ThemeAwareViewEngine.cs]: Please ensure that `enabledModules` contains all extensions

File: src/Orchard/Utility/Extensions/HttpRequestExtensions.cs (hunk 9 above)
Comment (sebastienros) [inline src/Orchard/Utility/Extensions/HttpRequestExtensions.cs]: Please remove it

File: src/Orchard/Environment/Configuration/AppLocations.cs [hunk 14]
```diff
@@ -0,0 +1,56 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Configuration;
+using System.Linq;
+
+namespace Orchard.Environment.Configuration {
+    public class AppLocations : IDependency {
```
Comment (sebastienros) [inline src/Orchard/Environment/Configuration/AppLocations.cs]: Make an interface. For consistency with other global services.

File: src/Orchard/Environment/Configuration/AppLocations.cs [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+        public string[] CoreLocations;
+        public string[] ModuleLocations;
+        public string[] ThemeLocations;
+        public string[] CommonLocations;  // locations that should not be common and not related to the current tenant
+        public string[] ModuleAndThemeLocations {  get { return ModuleLocations.Concat(ThemeLocations).Distinct(StringComparer.CurrentCultureIgnoreCase).ToArray(); } }
+
+        public AppLocations(IAppConfigurationAccessor appConfigurationAccessor)
```
Comment (sebastienros) [inline src/Orchard/Environment/Configuration/AppLocations.cs]: Remove

File: src/Orchard/Environment/Configuration/AppLocations.cs (hunk 14 above)
Comment (sebastienros) [inline src/Orchard/Environment/Configuration/AppLocations.cs]: IExtensionLocations

File: src/Orchard/Environment/OrchardStarter.cs [hunk 16]
```diff
@@ -35,11 +35,19 @@
 using Orchard.Services;
 using Orchard.WebApi;
 using Orchard.WebApi.Filters;
+using System.Linq;
+using System.Web.Configuration;
 
 namespace Orchard.Environment {
     public static class OrchardStarter {
         public static IContainer CreateHostContainer(Action<ContainerBuilder> registrations) {
+            AppLocations applLocations = new AppLocations(null);
```
Comment (sebastienros) [inline src/Orchard/Environment/OrchardStarter.cs]: builder.RegisterType<DefaultExtensionLocations>().As<IExtensionLocations>().SingleInstance()

File: src/Orchard/Mvc/ViewEngines/ThemeAwareness/ThemeAwareViewEngine.cs (hunk 13 above)
Comment (qt1) [inline src/Orchard/Mvc/ViewEngines/ThemeAwareness/ThemeAwareViewEngine.cs]: Seems like the original code had a minor bug: enabledModules is duplicated by concatination then filtered by distinct.. other than that all seems to be OK.

**Comment by sebastienros** [comment]
Can you check the merge conflicts?

**Comment by qt1** [comment]
merged with most recent 1.9.x

**Comment by sebastienros** [comment]
I am reluctant to include it in 1.9.2 directly, it can potentially break a whole site if a module is incorrectly loaded, or other modules are loaded where they weren't. We'll talk about it tomorrow during the triage. Otherwise, great work!

**Comment by qt1** [comment]
Tomorrow I am planning to attend a [slightly different version of a triage](https://www.youtube.com/watch?v=uGfoFA4is_0) namely [CV2 live](http://infected-mushroom.com/video/) so I can't participate in the discussion. 
To add confusion please see #5768 for an alternative that can be merged on dev instead.

**Comment by Piedone** [comment]
@qt1 The other PR is closed. Please open the one that you want to have merged and close the other one.

**Comment by qt1** [comment]
I have prepared two options: This one - base on 1.9.x - which is a bit cleaner (and I am using it internally) and I prefer it - but would effect 1.9.2 so it should be tested more people for possible regressions;
The one base on dev would not effect 1.9.2 but has some extra files that probably got in due to synchronization and changes in dev/1.9.x that where made parallel to the preparation and related conflicts.

At the moment I am leaving the discussion about selecting variants to the orchard gods, and I am going to an infected concert instead :)

**Comment by sebastienros** [comment]
So where is the actual PR for dev ?

**Comment by qt1** [comment]
I'll prepare a new one, hopefully in a couple of days.

**Comment by qt1** [comment]
updated against current 1.9.x.

File: src/Orchard/Environment/Configuration/ExtensionLocations.cs [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+    public class ExtensionLocations : IDependency {
+        public string[] CoreLocations;
+        public string[] ModuleLocations;
+        public string[] ThemeLocations;
+        public string[] CommonLocations;  // locations that should not be common and not related to the current tenant
+        public string[] ModuleAndThemeLocations;
+        public string[] ExtensionsVirtualPathPrefixes;  // Modules+Themes (no core)
+
+        public ExtensionLocations()
+        {
```
Comment (sebastienros) [inline src/Orchard/Environment/Configuration/ExtensionLocations.cs]: Please fix the formatting on the whole file.

Comment (qt1) [inline src/Orchard/Environment/Configuration/ExtensionLocations.cs]: formatting fixed. updated to current 1.9.x (no conflicts)

**Comment by sebastienros** [comment]
Now you can redo it and actually target __dev__ as we discussed a long time ago ;)

**Comment by qt1** [comment]
Please find yet another a PR against dev #5973

I would recommend to use the current PR # 5604 (this PR) since 1.9.2 is already on a separate branch 
and my tests where on the 1.9.x - so it should be better tested.

Please note Dev still has problems compiling (takes several rounds).

**Comment by sebastienros** [comment]
Why shouldn't we use the dev one? It can just be better.

**Comment by qt1** [comment]
My working branch is base on 1.9.x . I already have a few modules arranged in a separate directory using this PR. My personal testing and fixes will be done relative to 1.9.x.

Preparing an additional dev variant is double work for me so naturally I would prefer the PR be integrated to 1.9.x now.
In future will probably also be merged into dev next time 1.9.x is merged into dev.

BTW As far as I can tell PR is not breaking anything, so with 1.9.2 on a separate branch integrating it with 1.9.x seems like a low risk.

Please note the squashed variant is #5982
